keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,145 @@
+
+#### iptables NOTRACK
+
+NOTRACK rules are added for connections to and from the nodelocal dns service ip. Additional rules in FILTER table to whitelist these connections, since the INPUT and OUTPUT chains have a default DROP policy.
+
+A dnscache nanny, similar to the [dnsmasq nanny](https://github.com/kubernetes/dns/tree/master/pkg/dnsmasq) will create the dummy interface and iptables rules. The nanny gets the nodelocal dns service ip address by querying kube-dns for the service name. The nanny will also periodically ensure that the iptables rules are present. The resource usage for periodic iptables check needs to be measured. We can reduce the frequency of the checks by having the nanny query the dns cache and checking/adding rules only if the query fails.
+
+
+[Proposal presentation](https://docs.google.com/presentation/d/1c43cZqbVhGAlw3dSNQIOGuvQmDfKaA2yiAPRoYpa6iY), also shared at the sig-networking meeting on 2018-10-04
+
+Slide 5 has a diagram showing how the new dns cache fits in.
+
+#### Choice of caching agent
+
+The current plan is to run [Unbound dns server](https://www.nlnetlabs.nl/projects/unbound/about/) by default, based on these benchmark[ tests](https://github.com/kubernetes/perf-tests/tree/master/dns).
+
+Will rerun benchmark tests with coreDNS after SO_REUSEPORT support and measure QPS.
+
+Tests were run on a 1.9.7 cluster with 2 nodes, using Unbound 1.7.3 and coreDNS 1.2. Unbound QPS was 3600(caching)/3400(no caching), coreDNS 950(caching)/500(no caching)
+
+
+Based on the prototype/test results, these are the recommended defaults: 
+CPU request: 50m
+Memory Limit : 25m  
+Resource usage and QPS to be recomputed after metrics collection has been added.
+
+#### Metrics
+
+Unbound metrics can be extracted from the program unbound-control. This can be run from the nanny container.
+Unbound-control-setup is required to setup keys/certs for SSL.
+
+
+### Risks and Mitigations
+
+* Having the pods query the nodelocal cache introduces a single point of failure.   
+This is mitigated by assigning a PriorityClass of "system-node-critical", so it is always running. The nanny process will periodically check if the dns agent is running.
+Populating both the nodelocal cache ip address and kube-dns ip address in resolv.conf is not a reliable option. Depending on underlying implementation, this can result in kube-dns being queried only if cache ip does not repond, or both queried simultaneously.
+
+
+## Graduation Criteria
+TODO
+
+## Rollout Plan
+This feature will be launched with Alpha support in the first release. Master versions v1.13 and above will deploy the new add-on. Node versions v1.13 and above will have kubelet code to modify pods' resolv.conf. Nodes running older versions will run the nodelocal daemonset, but it will not be used. The user can specify a custom dnsConfig to use this local cache dns server.
```

Comment:
This seems to be more a deployment-tool choice, doesn't it?  E.g. are we saying that Kubeadm will support it?

Reply:
Likely not in the first release, but can be added later. I see Kubeadm supports kube-dns, coreDNS add-ons.

Reply:
That what does it mean to have a rollout plan?  If kubeadm doesn't support it, it's just a specification of how a user could set up their DNS, right?

Reply:
Not sure I understand. Kubeadm is one of the ways to bring up a cluster, right? You mean for users of kubeadm, this feature will just be a spec without actual support to run it? I can add the manifests for kubeadm to run this addon, it just might not happen by 1.13 code freeze.

Reply:
Tim's point is that deployment of each addon is specific to whatever is being used to create/manage the cluster. We have the same problem in making CoreDNS the "default". It requires adoption and updates to each of the cluster provisioning tools, not really to core Kubernetes itself. So, kubeadm, kops, kube-spray, etc. For CoreDNS, @rajansandeep updated several of these, I believe kube-up, kubeadm, and kops. The community followed with the rest.

Reply:
Exactly.  It's worthwhile to think through how someone would do this rollout, but all we can produce is a doc.